Code repository map (#3867)
numba/numba: docs/source/developer/repomap.rst

A Map of the Numba Repository
=============================

The Numba repository is quite large, and due to age has functionality spread
around many locations.  To help orient developers, this document will try to
summarize where different categories of functionality can be found.

.. note::
    It is likely that the organization of the code base will change in the
    future to improve organization.  Follow `issue #3807 <https://github.com/numba/numba/issues/3807>`_
    for more details.

Support Files
-------------

Build and Packaging
'''''''''''''''''''

- ``setup.py`` - Standard Python distutils/setuptools script
- ``MANIFEST.in`` - Distutils packaging instructions
- ``requirements.txt`` - Pip package requirements, not used by conda
- ``versioneer.py`` - Handles automatic setting of version in installed package from git tags

> - ``.flake8`` - Preferences for code formatting.  Files should be fixed and removed from exception list as time allows.

Reviewer: ```suggestion
- ``.flake8`` - Preferences for code formatting.  Files should be fixed and removed from the exception list as time allows.
```

- ``buildscripts/condarecipe.local`` - Conda build recipe

> - ``buildscripts/remove_unwanted_files.py`` - Helper script to clean up files.  Used by build recipes.

Reviewer: ```suggestion
- ``buildscripts/remove_unwanted_files.py`` - Helper script to remove files that will not compiled under Python 2.  Used by build recipes.
```

> - ``buildscripts/condarecipe_clone_icc_rt`` - Recipe to build a standalone icc_rt package.  Not used anymore?

Reviewer: May need regular re-runs against newer ICC RT's from Intel? There's no `icc_rt` package on defaults for non-Windows platforms (Windows has it due to Fortran? compilers).

Reviewer: In fact, we do, there was a patch to update last week!

Continuous Integration
''''''''''''''''''''''
- ``.binstar.yml`` - Binstar Build CI config (inactive)
- ``azure-pipelines.yml`` - Azure Pipelines CI config (active: Win/Mac/Linux)
- ``buildscripts/azure/`` - Azure Pipeline configuration for specific platforms
- ``.travis.yml`` - Travis CI config (active: Mac/Linux, will be dropped in the future)
- ``appveyor.yml`` - Appveyor CI config (inactive: Win)
- ``buildscripts/appveyor/`` - Appveyor build scripts
- ``buildscripts/incremental/`` - Generic scripts for building Numba on various CI systems
- ``codecov.yml`` - Codecov.io coverage reporting

Documentation / Examples
''''''''''''''''''''''''
- ``LICENSE`` - License for Numba
- ``LICENSES.third-party`` - License for third party code vendored into Numba
- ``README.rst`` - README for repo, also uploaded to PyPI

> - ``CONTRIBUTING.md`` - Documentation on how to contribute to project (out of date, should remove in favor of docs)

Reviewer: I think GH has a preference for this doc to exist, so suggest it is kept and forwards to docs?

- ``AUTHORS`` - List of Github users who have contributed PRs (out of date)
- ``CHANGE_LOG`` - History of Numba releases, also directly embedded into Sphinx documentation
- ``docs/`` - Documentation source
- ``docs/_templates/`` - Directory for templates (to override defaults with Sphinx theme)
- ``docs/Makefile`` - Used to build Sphinx docs with ``make``
- ``docs/source`` - ReST source for Numba documentation
- ``docs/_static/`` - Static CSS and image assets for Numba docs
- ``docs/gh-pages.py`` - Utility script to update Numba docs (stored as gh-pages)
- ``docs/make.bat`` - Not used (remove?)
- ``examples/`` - Example scripts demonstrating numba (re/move to numba-examples repo?)
- ``examples/notebooks/`` - Example notebooks (re/move to numba-examples repo?)
- ``benchmarks/`` - Benchmark scripts (re/move to numba-examples repo?)
- ``tutorials/`` - Tutorial notebooks (definitely out of date, should remove and direct to external tutorials)

> - ``numba/scripts/generate_lower_listing.py`` - Dump all registered lowering implementations for reference documentation

Reviewer: IIRC this dumps everything which is `@lower*` decorated, so misses all the lowering that occurs via the extending API.

Numba Source Code
-----------------

Numba ships with both the source code and tests in one package.

- ``numba/`` - all of the source code and tests

Public API
''''''''''

These define aspects of the public Numba interface.

- ``numba/decorators.py`` - User-facing decorators for compiling regular functions on the CPU
- ``numba/extending.py`` - Public decorators for extending Numba (``overload``, ``intrinsic``, etc)

> - ``numba/ccallback.py`` - ``@cfunc`` decorator for compiling functions to a fixed C singature.  Used to make callbacks.

Reviewer: ```suggestion
- ``numba/ccallback.py`` - ``@cfunc`` decorator for compiling functions to a fixed C signature.  Used to make callbacks.
```

Author: fixed

- ``numba/_dispatcher.{h,c}`` - C interface to C++ dispatcher implementation
- ``numba/_dispatcherimpl.cpp`` - C++ dispatcher implementation (for speed on common data types)

Compiler Pipeline
'''''''''''''''''

- ``numba/compiler.py`` - Compiler pipelines and flags
- ``numba/errors.py`` - Numba exception and warning classes 
- ``numba/ir.py`` - Numba IR data structure objects
- ``numba/bytecode.py`` - Bytecode parsing and function identity (??)
- ``numba/interpreter.py`` - Translate Python interpreter bytecode to Numba IR
- ``numba/analysis.py`` - Utility functions to analyze Numba IR (variable lifetime, prune branches, etc)
- ``numba/dataflow.py`` - Dataflow analysis for Python bytecode (used in analysis.py)
- ``numba/controlflow.py`` - Control flow analysis of Numba IR and Python bytecode
- ``numba/typeinfer.py`` - Type inference algorithm
- ``numba/transforms.py`` - Numba IR transformations
- ``numba/rewrites`` - Rewrite passes used by compiler
- ``numba/rewrites/__init__.py`` - Loads all rewrite passes so they are put into the registry

> - ``numba/rewrites/registry.py`` - Registry object for collecting rewrit passes

Reviewer: ```suggestion
- ``numba/rewrites/registry.py`` - Registry object for collecting rewrite passes
```

- ``numba/rewrites/ir_print.py`` - Write print() calls into special print nodes in the IR

> - ``numba/rewrites/static_raise.py`` - Convert exceptions with static arguments into special form that can be lowered

Reviewer: ```suggestion
- ``numba/rewrites/static_raise.py`` - Converts exceptions with static arguments into a special form that can be lowered
```

Author: fixed

- ``numba/rewrites/macros.py`` - Generic support for macro expansion in the Numba IR

> - ``numba/rewrites/static_getitem.py`` - Rewrite getitem and setitem with constant arguments to allow type inference

Reviewer: ```suggestion
- ``numba/rewrites/static_getitem.py`` - Rewrites getitem and setitem with constant arguments to allow type inference
```

Author: fixed

> - ``numba/rewrites/static_binop.py`` - Rewrite binary operations (specifically ``**``) with constant arguments so faster code can be generated

Reviewer: ```suggestion
- ``numba/rewrites/static_binop.py`` - Rewrites binary operations (specifically ``**``) with constant arguments so faster code can be generated
```

Author: fixed

> - ``numba/inline_closurecall.py`` - Inline body of closure functions to call site.

Reviewer: IIRC this also contains code for the "array comprehensions" feature and both reduction inlining and stencil inlining?

Author: Added this information to description

- ``numba/pylowering.py`` - Lowering of Numba IR in object mode
- ``numba/pythonapi.py`` - LLVM IR code generation to interface with CPython API

Type Management
'''''''''''''''
- ``numba/typeconv/`` - Implementation of type casting and type signature matching in both C++ and Python
- ``numba/capsulethunk.h`` - Used by typeconv
- ``numba/types/`` - definition of the Numba type hierarchy, used everywhere in compiler to select implementations
- ``numba/consts.py`` - Constant inference (used to make constant values available during codegen when possible)
- ``numba/datamodel`` - LLVM IR representations of data types in different contexts
- ``numba/datamodel/models.py`` - Models for most standard types
- ``numba/datamodel/registry.py`` - Decorator to register new data models
- ``numba/datamodel/packer.py`` - Pack typed values into a data structure
- ``numba/datamodel/testing.py`` - Data model tests (this should move??)
- ``numba/datamodel/manager.py`` - Map types to data models

Compiled Extensions
'''''''''''''''''''

Numba uses a small amount of compiled C/C++ code for core functionality, like

> dispatching and type matching where performance matters, and it is more

Reviewer: ```suggestion
dispatching and type matching, where performance matters, and it is more
```

> convient to encapsulate direct interaction with CPython APIs.

Reviewer 2: convenient to encapsulate direct interaction with CPython APIs.
```suggestion
convient to encapsulate direct interaction with CPython APIs.
```

> - ``numba/_arraystruct.h`` - Struct for holding NumPy array attributes.  Used in helperlib and nrt.

Reviewer: `nrt` is not defined anywhere.

Author: Switched to "Numba Runtime"

- ``numba/_helperlib.c`` - C functions required by Numba compiled code at runtime.  Linked into ahead-of-time compiled modules
- ``numba/_helpermod.c`` - Python extension module with pointers to functions from ``_helperlib.c`` and ``_npymath_exports.c``
- ``numba/_npymath_exports.c`` - Export function pointer table to NumPy C math functions
- ``numba/_dynfuncmod.c`` - Python extension module exporting _dynfunc.c functionality
- ``numba/_dynfunc.c`` - C level Environment and Closure objects (keep in sync with numba/target/base.py)
- ``numba/mathnames.h`` - Macros for defining names of math functions
- ``numba/_pymodule.h`` - C macros for Python 2/3 portable naming of C API functions
- ``numba/_math_c99.{h,c}`` - C99 math compatibility (needed Python 2.7 on Windows, compiled with VS2008)

> - ``numba/mviewbuf.c`` - Handles Python memorviews

Reviewer: ```suggestion
- ``numba/mviewbuf.c`` - Handles Python memoryviews
```

Author: fixed

- ``numba/_typeof.{h,c}`` - C implementation of type fingerprinting, used by dispatcher
- ``numba/_numba_common.h`` - Portable C macro for marking symbols that can be shared between object files, but not outside the library.

Misc Support
''''''''''''

- ``numba/_version.py`` - Updated by versioneer

> - ``numba/runtime`` - Global singleton that manages memory allocated on the heap by Numba-compiled functions

Reviewer 3: Let's call out that it's the "language runtime".  Though all it does now is memory and reference-counting.

- ``numba/ir_utils.py`` - Utility functions for working with Numba IR data structures 
- ``numba/cgutils.py`` - Utility functions for generating common code patterns in LLVM IR
- ``numba/six.py`` - Vendored subset of ``six`` package for Python 2 + 3 compatibility
- ``numba/io_support.py`` - Workaround for various names of StringIO in different Python versions (should this be in six?)
- ``numba/utils.py`` - Python 2 backports of Python 3 functionality (also imports local copy of ``six``)
- ``numba/appdirs.py`` - Vendored package for determining application config directories on every platform

> - ``numba/compiler_lock.py`` - Global compiler lock because our usage of LLVM is not thread-safe

Reviewer: ```suggestion
- ``numba/compiler_lock.py`` - Global compiler lock because Numba's use of LLVM is not thread-safe
```

Author: fixed

- ``numba/special.py`` - Python stub implementations of special Numba functions (prange, gdb*)

> - ``numba/servicelib`` - Should be removed?

Reviewer 3: The `TLSStack` is used by the cuda and roc backend for managing the device stack.

- ``numba/itanium_mangler.py`` - Python implementation of Itanium C++ name mangling
- ``numba/findlib.py`` - Helper function for locating shared libraries on all platforms
- ``numba/debuginfo.py`` - Helper functions to construct LLVM IR debug info
- ``numba/unsafe`` - ``@intrinsic`` helper functions that can be used to implement direct memory/pointer manipulation from nopython mode functions
- ``numba/unsafe/refcount.py`` - Read reference count of object
- ``numba/unsafe/tuple.py`` - Replace a value in a tuple slot
- ``numba/unsafe/ndarray.py`` - NumPy array helpers
- ``numba/unsafe/bytes.py`` - Copying and dereferencing data from void pointers

> - ``numba/dummyarray.py`` - ???

Reviewer: This is used through composition by the GPU backends in the device array classes so as to have an on host mock-array representation of the device array which holds all the array info but not the actual data.

Author: Summarized this description

- ``numba/callwrapper.py`` - Handles argument unboxing and releasing the GIL when moving from Python to nopython mode
- ``numba/ctypes_support.py`` - Import this instead of ``ctypes`` to workaround portability issue with Python 2.7
- ``numba/cffi_support.py`` - Alias of numba.typing.cffi_utils for backward compatibility (still needed?)

> - ``numba/numpy_support.py`` - Helper functions for working with NumPy and dtypes

Reviewer: also Numba type interoperability with NumPy.

Author: rephrased to make this clear

- ``numba/tracing.py`` - Decorator for tracing Python calls and emitting log messages

> - ``numba/funcdesc.py`` - Classes for describing function metadata (used in compiler)

Reviewer: ```suggestion
- ``numba/funcdesc.py`` - Classes for describing function metadata (used in the compiler)
```

Author: fixed

- ``numba/sigutils.py`` - Helper functions for parsing and normalizing Numba type signatures
- ``numba/serialize.py`` - Support for pickling compiled functions
- ``numba/caching.py`` - Disk cache for compiled functions
- ``numba/npdatetime.py`` - Helper functions for implementing NumPy datetime64 support

Core Python Data Types
''''''''''''''''''''''

> - ``numba/_hashtable.{h,c}`` - Adaptation of Python 3.7 hash table implementation

Reviewer: ```suggestion
- ``numba/_hashtable.{h,c}`` - Adaptation of the Python 3.7 hash table implementation
```

Author: fixed

- ``numba/_dictobject.{h,c}`` - C level implementation of typed dictionary 
- ``numba/dictobject.py`` - Nopython mode wrapper for typed dictionary
- ``numba/unicode.py`` - Unicode strings (Python 3.5 and later)
- ``numba/typed`` - Python interfaces to statically typed containers
- ``numba/typed/typeddict.py`` - Python interface to typed dictionary
- ``numba/jitclass`` - Implementation of JIT compilation of Python classes
- ``numba/generators.py`` - Support for lowering Python generators

Math
''''

- ``numba/_random.c`` - Reimplementation of NumPy / CPython random number generator

> - ``numba/_lapack.c`` - Wrappers for calling BLAS functions

Reviewer: ```suggestion
- ``numba/_lapack.c`` - Wrappers for calling BLAS and LAPACK functions (requires SciPy)
```

Author: fixed

ParallelAccelerator
'''''''''''''''''''

> Code transformation passes that extract parallelizable code from function and convert to multithreaded gufunc calls.

Reviewer: ```suggestion
Code transformation passes that extract parallelizable code from a function and convert it into multithreaded gufunc calls.
```

Author: fixed

- ``numba/targets/randomimpl.py`` - Python and NumPy ``random`` modules
- ``numba/targets/npyimpl.py`` - Implementations of most NumPy ufuncs
- ``numba/targets/slicing.py`` - Slice objects, and index calculations used in slicing
- ``numba/targets/numbers.py`` - Numeric values (int, float, etc)
- ``numba/targets/listobj.py`` - Python lists
- ``numba/targets/fastmathpass.py`` - Rewrite pass to add fastmath attributes to function call sites and binary operations
- ``numba/targets/removerefctpass.py`` - Rewrite pass to remove unnecessary incref/decref pairs
- ``numba/targets/cffiimpl.py`` - CFFI functions
- ``numba/targets/descriptors.py`` - empty base class for all target descriptors (is this needed?)
- ``numba/targets/arraymath.py`` - Math operations on arrays (both Python and NumPy)
- ``numba/targets/linalg.py`` - NumPy linear algebra operations
- ``numba/targets/rangeobj.py`` - Python `range` objects
- ``numba/targets/npyfuncs.py`` - Kernels used in generating some NumPy ufuncs
- ``numba/targets/arrayobj.py`` - Array operations (both NumPy and buffer protocol)
- ``numba/targets/enumimpl.py`` - Enum objects

> - ``numba/targets/polynomial.py`` - ``numpy.roots`` function

Reviewer: Should probably implement the rest of the polynomial function support. Note to self. Write tickets.

- ``numba/npyufunc/parfor.py`` - gufunc lowering for ParallelAccelerator
- ``numba/npyufunc/parallel.py`` - Codegen for ``parallel`` target
- ``numba/npyufunc/array_exprs.py`` - Rewrite pass for turning array expressions in regular functions into ufuncs
- ``numba/npyufunc/wrappers.py`` - Wrap scalar function kernel with loops
- ``numba/npyufunc/workqueue.{h,c}`` - Threading backend based on pthreads/Windows threads and queues
- ``numba/npyufunc/omppool.cpp`` - Threading backend based on OpenMP
- ``numba/npyufunc/tbbpool.cpp`` - Threading backend based on TBB

Unit Tests (CPU)
''''''''''''''''

CPU unit tests (GPU target unit tests listed in later sections

- ``runtests.py`` - Convenience script that launches test runner and turns on full compiler tracebacks
- ``run_coverage.py`` - Runs test suite with coverage tracking enabled
- ``.coveragerc`` - Coverage.py configuration
- ``numba/runtests.py`` - Entry point to unittest runner
- ``numba/_runtests.py`` - Implementation of custom test runner command line interface
- ``numba/tests/test_*`` - Test cases
- ``numba/tests/*_usecases.py`` - Python functions compiled by some unit tests

> - ``numba/tests/support.py`` - Helper functions for testin and special TestCase implementation

Reviewer: ```suggestion
- ``numba/tests/support.py`` - Helper functions for testing and special TestCase implementation
```

- ``numba/tests/dummy_module.py`` - Module used in ``test_dispatcher.py``
- ``numba/tests/npyufunc`` - ufunc / gufunc compiler tests

> - ``numba/unittest_support.py`` - Import instead of unittest to handle portability issues (no longer needed?)

Reviewer: seems like it might not be needed.

- ``numba/cuda/target.py`` - Typing and target contexts for GPU
- ``numba/cuda/cudamath.py`` - Type signatures for math functions in CUDA Python
- ``numba/cuda/errors.py`` - Validation of kernel launch configuration
- ``numba/cuda/nvvmutils.py`` - Helper functions for generating NVVM-specific IR
- ``numba/cuda/testing.py`` - Support code for creating CUDA unit tests and capturing standard out
- ``numba/cuda/cudadecl.py`` - Type signatures of CUDA API (threadIdx, blockIdx, atomics) in Python on GPU
- ``numba/cuda/cudaimpl.py`` - Implementations of CUDA API functions on GPU
- ``numba/cuda/codegen.py`` - Code generator object for CUDA target
- ``numba/cuda/cudadrv/`` - Wrapper around CUDA driver API
- ``numba/cuda/tests/`` - CUDA unit tests, skipped when CUDA is not detected
- ``numba/cuda/tests/cudasim/`` - Tests of CUDA simulator
- ``numba/cuda/tests/nocuda/`` - Tests for NVVM functionality when CUDA not present
- ``numba/cuda/tests/cudapy/`` - Tests of compiling Python functions for GPU
- ``numba/cuda/tests/cudadrv/`` - Tests of Python wrapper around CUDA API

ROCm GPU Target
'''''''''''''''

> Note that the ROCm target does reuse some parts of the CPU target, but

Reviewer: ```suggestion
Note that the ROCm target does reuse some parts of the CPU target, and
```

duplicates some code from CUDA target.  A future refactoring could pull out
the common subset of CUDA and ROCm.  An older version of this target was based
on the HSA API, so "hsa" appears in many places.

- ``numba/roc`` - ROCm GPU target for AMD GPUs
- ``numba/roc/descriptor.py`` - TargetDescriptor subclass for ROCm target
- ``numba/roc/enums.py`` - Internal constants
- ``numba/roc/mathdecl.py`` - Declarations of math functions that can be used on device
- ``numba/roc/mathimpl.py`` - Implementations of math functions for device
- ``numba/roc/compiler.py`` - Compiler pipeline for ROCm target
- ``numba/roc/hlc`` - Wrapper around LLVM interface for AMD GPU
- ``numba/roc/initialize.py`` - Register ROCm target for ufunc/gufunc compiler
- ``numba/roc/hsadecl.py`` - Type signatures for ROCm device API in Python
- ``numba/roc/hsaimpl.py`` - Implementations of ROCm device API
- ``numba/roc/dispatch.py`` - ufunc/gufunc dispatcher

> - ``numba/roc/README.md`` - Notes on testing target (delete?)

Reviewer: think this can be deleted, not applicable in current code base
